### PR TITLE
Add support for /public

### LINKED ISSUE #7177: [RFC] Serving static files from the root
# Feature request

## Is your feature request related to a problem? Please describe.

After seeing some custom server.js usage a clear pattern is people using it to define static file serving from the root. For example for robots.txt or google site verification and other files that have to be served from the root.

## Describe the solution you'd like

Ideally users would not have to define this. The common way root serving has been solved is through a `public` directory.

For example:
- [Symfony](https://symfony.com/doc/current/configuration/override_dir_structure.html)
- [create-react-app](https://github.com/facebook/create-react-app#yarn)

### Notes

This also needs to be applied to `@now/next` when implemented.

### PR DESCRIPTION
**Pull request by lfades**
Closes #7177 

Implements the RFC described here: #7177.

Files in the `public` directory are automatically served from the root.
Pages are preferred though, so if you create:

- pages/about.js
- public/about

The `pages/about.js` will be rendered.

PR for `@now/next` can be found here: https://github.com/zeit/now-builders/pull/457

### CODE REVIEW

File: packages/next-server/server/next-server.ts [hunk 1]
```diff
@@ -266,15 +267,27 @@ export default class Server {
...
+      })
+      // Request was ended by the user
+      if (html === null) {
+        return
+      }
 
-    return this.sendHTML(req, res, html)
+      return this.sendHTML(req, res, html)
+    } catch (err) {
+      if (err && err.code === 'ENOENT') {
```
Comment (ijjk) [inline packages/next-server/server/next-server.ts]: This isn't reached in dev mode, will need to add a check in `next-dev-server` `renderToHTML`. Should probably add a test to make sure it's working in dev mode

Comment (lfades) [reply packages/next-server/server/next-server.ts]: @ijjk Done, thank you.

File: packages/next-server/server/next-server.ts [hunk 2]
```diff
@@ -93,8 +101,11 @@ export default class Server {
       publicRuntimeConfig,
     })
 
-    const routes = this.generateRoutes()
-    this.router = new Router(routes)
+    this.router = new Router()
+    this.routesPromise = this.generateRoutes().then((routes) => {
```
Comment (timneutkens) [inline packages/next-server/server/next-server.ts]: This actually makes Next.js serve 404 on bootup 🤔 It should be done in `prepare` imo.

File: packages/next-server/server/next-server.ts [hunk 3]
```diff
@@ -44,6 +49,7 @@ export default class Server {
     assetPrefix?: string,
   }
   router: Router
+  routesPromise?: Promise<void>
```
Comment (timneutkens) [inline packages/next-server/server/next-server.ts]: I think this is leftover from the previous implementation you did 🤔

File: packages/next-server/server/next-server.ts [hunk 4]
```diff
@@ -214,11 +227,36 @@ export default class Server {
...
+
+    return routes
+  }
+
   private async run(
     req: IncomingMessage,
     res: ServerResponse,
     parsedUrl: UrlWithParsedQuery,
   ) {
+    if (this.routesPromise) await this.routesPromise
```
Comment (timneutkens) [inline packages/next-server/server/next-server.ts]: Same as https://github.com/zeit/next.js/pull/7213/files#r280705505

**Review by timneutkens — APPROVED** [review]

**Review by ijjk — APPROVED** [review]

**Comment by pozylon** [comment]
Just tested this and it does not seem to work as described. Added a public folder to my project and put a robots.txt there. getting 404 if navigating to http://localhost:3000/robots.txt.

I've tried to remove .next folder, use next build & start instead of dev, no success...

**Comment by pozylon** [comment]
Found out it's behind an experimental flag "publicDirectory", all good

**Comment by jaydenseric** [comment]
For other people wondering, here are instructions to enable the experimental feature: https://spectrum.chat/next-js/general/clarification-on-public-directory~67b410b6-81ff-442b-8238-3a6f72891384?m=MTU2MjY0NjQ2NjI4NA==

**Pull request merged**